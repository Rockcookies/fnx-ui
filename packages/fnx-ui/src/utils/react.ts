import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
