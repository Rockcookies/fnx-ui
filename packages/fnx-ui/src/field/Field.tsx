import React, {
	cloneElement,
	forwardRef,
	isValidElement,
	useContext,
} from 'react';
import Cell from '../cell';
import { FormHelpAlign, FormLabelAlign } from '../form';
import { FormContext, FormItemContext } from '../form/context';
import configComponentProps from '../hooks/config-component-props';
import useMergedProp from '../hooks/use-merged-prop';
import { addUnit } from '../utils/format';
import { Dictionary } from '../utils/interface';
import { classnames, createBEM } from '../utils/namespace';
import { FieldProps } from './interface';

const NS = 'fnx-field';
const bem = createBEM(NS);

const useProps = configComponentProps<
	Required<
		Pick<
			FieldProps,
			| 'labelProps'
			| 'controlProps'
			| 'titleProps'
			| 'contentProps'
			| 'disabled'
			| 'help'
		>
	>
>({
	labelProps: {},
	controlProps: {},
	titleProps: {},
	contentProps: {},
	disabled: false,
	help: [],
});

const Field = forwardRef<HTMLDivElement, FieldProps>((_props, ref) => {
	const [
		{ labelProps, controlProps, titleProps, contentProps, disabled, help },
		{
			labelAlign: _labelAlign,
			helpAlign: _helpAlign,
			colon: _colon,
			required: _required,
			labelWidth: _labelWidth,
			label,
			controlPrefix,
			controlSuffix,
			className,
			children,
			...restProps
		},
	] = useProps(_props);

	const formContext = useContext(FormContext);
	const { meta, isRequired: fieldIsRequired } = useContext(FormItemContext);

	const labelAlign = useMergedProp<FormLabelAlign>(
		'left',
		_labelAlign,
		formContext.labelAlign,
	);

	const helpAlign = useMergedProp<FormHelpAlign>(
		'left',
		_helpAlign,
		formContext.helpAlign,
	);
	const colon = useMergedProp<boolean>(false, _colon, formContext.colon);
	const required = useMergedProp<boolean>(
		formContext.requiredMark === true,
		_required,
	);

	const labelWidth = useMergedProp<string | number | undefined>(
		undefined,
		_labelWidth,
		formContext.labelWidth,
	);

	let helps = (meta && meta.errors) || [];

	if (Array.isArray(help)) {
		helps = [...help, ...helps];
	} else {
		helps = [help, ...helps];
	}

	const classes: Dictionary = {};

	if (meta && meta.validating) {
		classes.validating = true;
	} else if (helps.length > 0) {
		classes.error = true;
	} else if (meta && meta.touched) {
		classes.success = true;
	}

	const renderChildren = () => {
		if (!isValidElement(children)) {
			return children;
		}

		if (children.props.disabled != null) {
			return children;
		}

		return cloneElement(children, {
			...children.props,
			disabled,
		});
	};

	return (
		<Cell
			{...restProps}
			className={classnames(
				bem([
					{
						...classes,
						required:
							fieldIsRequired != null
								? fieldIsRequired
								: required,
						disabled,
					},
				]),
				className,
			)}
			ref={ref}
			title={label}
			titleProps={{
				...labelProps,
				className: classnames(
					bem('label', [`text-${labelAlign}`, { colon }]),
					labelProps.className,
					titleProps.className,
				),
				style: {
					...(labelWidth !== 0 ? { width: addUnit(labelWidth) } : {}),
					...titleProps.style,
				},
			}}
			contentProps={{
				...controlProps,
				...contentProps,
				className: classnames(
					bem('control'),
					controlProps.className,
					contentProps.className,
				),
			}}
		>
			<div className={bem('control-body')}>
				{controlPrefix && (
					<div className={bem('control-prefix')}>{controlPrefix}</div>
				)}
				<div className={bem('control-item')}>{renderChildren()}</div>
				{controlSuffix && (
					<div className={bem('control-suffix')}>{controlSuffix}</div>
				)}
			</div>
			{helps.length > 0 && (
				<div className={bem('help', [`text-${helpAlign}`])}>
					{helps.map((err, idx) => (
						<div key={idx}>{err}</div>
					))}
				</div>
			)}
		</Cell>
	);
});

Field.displayName = 'Field';

export default Field;
