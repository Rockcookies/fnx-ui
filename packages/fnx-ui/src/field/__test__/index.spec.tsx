import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import TestsDOM from '../../../test/dom';
import Field from '../index';

describe('<Form.Item/>', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	/**----------------------FieldInput---------------------- */

	it('number type', () => {
		const onChange = jest.fn();

		const { container } = render(
			<Field label={'number'}>
				<Field.Input
					type="number"
					defaultValue="0"
					onChange={onChange}
				/>
			</Field>,
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'input',
		);

		expect(input.value).toEqual('0');

		fireEvent.change(input, { target: { value: '1.4.' } });
		expect(input.value).toEqual('1.4');

		fireEvent.change(input, { target: { value: '123abc' } });
		expect(input.value).toEqual('123');
	});

	it('digit type', () => {
		const { container } = render(
			<Field label={'digit'}>
				<Field.Input type="digit" defaultValue="0" />
			</Field>,
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'input',
		);

		expect(input.value).toEqual('0');

		fireEvent.change(input, { target: { value: '1.' } });
		expect(input.value).toEqual('1');

		fireEvent.change(input, { target: { value: '123abc' } });

		expect(input.value).toEqual('123');
	});

	it('trigger onChange、onFocus、onBlur', () => {
		const onFocus = jest.fn();
		const onBlur = jest.fn();
		const onChange = jest.fn();

		const { container } = render(
			<Field label={'digit'}>
				<Field.Input
					type="digit"
					defaultValue="1"
					onFocus={onFocus}
					onBlur={onBlur}
					onChange={onChange}
				/>
			</Field>,
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'input',
		);

		act(() => {
			fireEvent.focus(input);
		});
		expect(onFocus).toHaveBeenCalledWith('1');

		act(() => {
			fireEvent.blur(input);
			jest.advanceTimersByTime(500);
		});
		expect(onBlur).toHaveBeenCalledWith('1');

		fireEvent.change(input, { target: { value: '14' } });
		expect(onChange).toHaveBeenCalledWith('14');
	});

	it('clearable prop and clear-trigger', () => {
		const { container, rerender } = render(
			<Field>
				<Field.Input clearable={true} defaultValue="3" />
			</Field>,
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'input',
		);

		expect(container.querySelector('.fnx-field-input__clear')).toBeNull();

		fireEvent.focus(input);
		expect(
			container.querySelector('.fnx-field-input__clear'),
		).not.toBeNull();

		rerender(
			<Field>
				<Field.Input
					clearable={true}
					defaultValue="3"
					clearTrigger="always"
				/>
			</Field>,
		);

		expect(
			container.querySelector('.fnx-field-input__clear'),
		).not.toBeNull();
	});

	it('inputAlign、readonly、disabled props', () => {
		const { container, rerender } = render(
			<Field>
				<Field.Input inputAlign="center" disabled={true} />
			</Field>,
		);

		expect(
			TestsDOM.mustQuerySelector(container, '.fnx-field-input--disabled'),
		).not.toBeNull();

		rerender(
			<Field>
				<Field.Input inputAlign="right" readOnly={true} />
			</Field>,
		);

		expect(
			TestsDOM.mustQuerySelector(
				container,
				'.fnx-field-input--text-right',
			),
		).not.toBeNull();
		expect(
			TestsDOM.mustQuerySelector(
				container,
				'.fnx-field-input__control',
			).getAttribute('readonly'),
		).toBeDefined();
	});

	it('maxlength', () => {
		const { container } = render(
			<Field>
				<Field.Input maxLength={3} defaultValue="12345" />
			</Field>,
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'input',
		);

		expect(input.value).toEqual('123');

		fireEvent.change(input, { target: { value: '1234' } });
		expect(input.value).toEqual('123');

		// reset
		fireEvent.change(input, { target: { value: '123' } });
		expect(input.value).toEqual('123');

		fireEvent.compositionStart(input);

		fireEvent.change(input, { target: { value: '1234' } });
		expect(input.value).toEqual('1234');

		fireEvent.compositionEnd(input);
		expect(input.value).toEqual('123');

		fireEvent.change(input, { target: { value: '😂😂😂😂' } });
		expect(input.value).toEqual('😂😂😂');
	});

	/**----------------------FieldTextArea---------------------- */
	it('textArea trigger onChange、onFocus、onBlur', () => {
		const onFocus = jest.fn();
		const onBlur = jest.fn();
		const onChange = jest.fn();

		const { container } = render(
			<Field label={'textarea'}>
				<Field.TextArea
					defaultValue="1"
					onFocus={onFocus}
					onBlur={onBlur}
					onChange={onChange}
				/>
			</Field>,
		);

		const textarea = TestsDOM.mustQuerySelector<HTMLTextAreaElement>(
			container,
			'textarea',
		);

		act(() => {
			fireEvent.focus(textarea);
		});

		expect(onFocus).toHaveBeenCalledWith('1');

		act(() => {
			fireEvent.blur(textarea);
			jest.advanceTimersByTime(500);
		});

		expect(onBlur).toHaveBeenCalledWith('1');

		fireEvent.change(textarea, { target: { value: '14' } });
		expect(onChange).toHaveBeenCalledWith('14');
	});

	it('textArea maxlength and show count', () => {
		const { container } = render(
			<Field>
				<Field.TextArea
					maxLength={3}
					defaultValue="12345"
					showCount={true}
				/>
			</Field>,
		);

		const textarea = TestsDOM.mustQuerySelector<HTMLTextAreaElement>(
			container,
			'textarea',
		);

		const count = TestsDOM.mustQuerySelector(
			container,
			'.fnx-field-textarea__count',
		);

		expect(textarea.value).toEqual('123');

		fireEvent.change(textarea, { target: { value: '1234' } });
		expect(textarea.value).toEqual('123');
		expect(count.textContent).toEqual('3 / 3');

		fireEvent.change(textarea, { target: { value: '1234' } });
		expect(textarea.value).toEqual('123');

		// reset
		fireEvent.change(textarea, { target: { value: '123' } });
		expect(textarea.value).toEqual('123');

		fireEvent.compositionStart(textarea);

		fireEvent.change(textarea, { target: { value: '1234' } });
		expect(textarea.value).toEqual('1234');

		fireEvent.compositionEnd(textarea);
		expect(textarea.value).toEqual('123');

		fireEvent.change(textarea, { target: { value: '😂😂😂😂' } });
		expect(textarea.value).toEqual('😂😂😂');
	});

	/**----------------------Field---------------------- */
	it('label、labelAlign、labelWidth props', () => {
		const { container } = render(
			<Field label="field" labelAlign="right" labelWidth="44">
				<Field.Input />
			</Field>,
		);
		expect(container).toMatchSnapshot();
	});

	it('required prop', () => {
		const { container } = render(
			<Field label="field" required>
				<Field.Input />
			</Field>,
		);

		expect(container.querySelector('.fnx-field--required')).not.toBeNull();
	});

	it('help prop', () => {
		const { container } = render(
			<Field label="field" help="validate Error" helpAlign="right">
				<Field.Input />
			</Field>,
		);

		const help = TestsDOM.mustQuerySelector(
			container,
			'.fnx-field__help--text-right',
		);

		expect(help.textContent).toEqual('validate Error');
	});

	it('should autoSize textarea field', async () => {
		const { container } = render(
			<Field label="field">
				<Field.TextArea autoSize={{}} />
			</Field>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should allow autoSize prop be be an object', async () => {
		const { container } = render(
			<Field label="field">
				<Field.TextArea autoSize={{ minHeight: 50 }} />
			</Field>,
		);

		expect(container).toMatchSnapshot();
	});
});
