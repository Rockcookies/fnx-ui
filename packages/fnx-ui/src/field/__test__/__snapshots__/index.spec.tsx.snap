// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form.Item/> label、labelAlign、labelWidth props 1`] = `
<div>
  <div
    class="fnx-cell fnx-cell--align-top fnx-field"
  >
    <div
      class="fnx-cell__title fnx-field__label fnx-field__label--text-right"
      style="width: 44px;"
    >
      field
    </div>
    <div
      class="fnx-cell__content fnx-field__control"
    >
      <div
        class="fnx-field__control-body"
      >
        <div
          class="fnx-field__control-item"
        >
          <span
            class="fnx-field-input"
          >
            <input
              class="fnx-field-input__control"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Form.Item/> should allow autoSize prop be be an object 1`] = `
<div>
  <div
    class="fnx-cell fnx-cell--align-top fnx-field"
  >
    <div
      class="fnx-cell__title fnx-field__label fnx-field__label--text-left"
    >
      field
    </div>
    <div
      class="fnx-cell__content fnx-field__control"
    >
      <div
        class="fnx-field__control-body"
      >
        <div
          class="fnx-field__control-item"
        >
          <span
            class="fnx-field-textarea"
          >
            <textarea
              class="fnx-field-textarea__control"
              rows="1"
              style="min-height: 50px; height: 50px;"
            />
            <textarea
              aria-hidden="true"
              class="fnx-field-textarea__auto-size"
              readonly=""
              rows="1"
              tabindex="-1"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Form.Item/> should autoSize textarea field 1`] = `
<div>
  <div
    class="fnx-cell fnx-cell--align-top fnx-field"
  >
    <div
      class="fnx-cell__title fnx-field__label fnx-field__label--text-left"
    >
      field
    </div>
    <div
      class="fnx-cell__content fnx-field__control"
    >
      <div
        class="fnx-field__control-body"
      >
        <div
          class="fnx-field__control-item"
        >
          <span
            class="fnx-field-textarea"
          >
            <textarea
              class="fnx-field-textarea__control"
              rows="1"
              style="min-height: 0px; height: 0px;"
            />
            <textarea
              aria-hidden="true"
              class="fnx-field-textarea__auto-size"
              readonly=""
              rows="1"
              tabindex="-1"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
